### ISSUE TITLE
narrow() at a nonzero sub-stick offset fails layout pass (case A2 of #4828, unfixed by #4829)

### ISSUE THREAD

**Issue by dgrove-oss** [torch-spyre/torch-spyre#4834]
## Problem

`pool[ids.narrow(0, k, TILE)]` at a nonzero, sub-stick offset `k` fails to
compile with "no supported output layout found" -- with no `for_each_tile`
involved. This is case A2 from #4828's reproducer, and remains unfixed and
untested after #4829 (which fixed only case B, the `for_each_tile` compile
crash).

```
[A2 narrow offset 2] FAILED: Unsupported: Spyre backend does not support:
Multi-arg pointwise (buf0): no supported output layout found with
size=[2, 128], output_ea=ElementArrangement.STANDARD, coordinates=[d0, d1].
Input EAs: [<ElementArrangement.STANDARD: 0>, <ElementArrangement.STANDARD: 0>]
```

Per #4828's reproducer, `ids.narrow(0, 0, TILE)` (offset 0) compiles; the same
call at offset `TILE` (an int32 sub-stick offset, since 4 int32 `ids` share
one 128-byte/32-element stick) fails at every torch-spyre revision tried,
including current `main`.

## Relationship to #4828 / #4829

#4828 tracked three cases (A0/A2/B). #4829 fixed case B's compile crash and
added a refusal (`UnalignedStickSplit`) for a related sub-stick advance
inside `for_each_tile` gathers. Neither of #4829's changes touch case A2: it
has no `for_each_tile` splice, so `_check_indirect_index_step` in
`spyre_kernel.py` is never reached, and the failure is a plain layout-pass
"no supported output layout" rather than a raised `Unsupported`/
`UnalignedStickSplit`. The maintainer's second comment on #4828 explicitly
left this open: "Case A2 (narrow at offset 2) is still a separate compile-time
refusal. Whether it has the same underlying cause is not yet established."

## Reproducer

Case A2 from #4828's original reproducer:

```python
import torch
import torch_spyre  # noqa: F401  (registers the 'spyre' device)

POOL, WIDTH, TRIPS, TILE = 256, 128, 4, 2
DEV = torch.device("spyre")


def case_a_offset2(ids, pool, init):
    return init + pool[ids.narrow(0, TILE, TILE)]


ids = torch.arange(TRIPS, dtype=torch.int32).to(DEV)
pool = torch.randn(POOL, WIDTH, dtype=torch.float16).to(DEV)
init = torch.zeros(TILE, WIDTH, dtype=torch.float16).to(DEV)

torch.compile(case_a_offset2, dynamic=False, fullgraph=True)(ids, pool, init)
```

## Environment

- torch-spyre `main` (post #4829, `011e8df5`)
- torch `2.13.0+cpu`

## Next steps

Investigate whether this shares a root cause with the sub-stick
`for_each_tile` advance bug fixed/refused by #4829 (tracked separately in
#TODO), or is a distinct layout-pass gap for a plain `narrow` view with no
loop/indirect-index machinery involved.

**Comment by dgrove-oss** [comment]
Cross-reference: the sub-stick `for_each_tile` advance's device-test-coverage gap is tracked separately as #4835.

**Comment by dgrove-oss** [comment]
## Root cause

`ids.narrow(0, TILE, TILE)` fails because `_decompose_constant_offset`
(`torch_spyre/_inductor/views.py:205`) has no non-stick dimension available
to absorb the `narrow`'s constant offset when the indexed tensor's entire
content fits in a single device stick.

Traced with the reproducer from this issue's description
(`POOL, WIDTH, TRIPS, TILE = 256, 128, 4, 2`):

- `ids` (int32, 4 elements) is committed to device layout
  `device_size=[1, 64], stride_map=[4, 1]` before this op is reached --
  everything fits in one 64-int32-element stick, so its only "outer"
  dimension has `size=1`.
- `ids.narrow(0, 2, 2)` introduces a constant `+2` offset into the
  consuming pointwise op's `MemoryDep` index for `ids` (`d0 + 2`, confirmed
  via instrumentation of `propagate_layouts._multi_arg_pointwise_layouts`).
- `compute_coordinates` (`views.py`) peels constant offsets through
  `_decompose_constant_offset`, which -- by design -- distributes the
  offset across dimensions with `size[d] > 1`, specifically so the offset
  does *not* land on the stick coordinate. Its own docstring: "Positional
  peeling keeps the stick coordinate offset-free."
- For `ids`, the **only** dimension with `size > 1` is the stick dimension
  itself (the outer dim is squeezed to size 1 by the single-stick layout).
  The `size[d] > 1` filter (`views.py:231`) excludes the outer dim, so the
  offset has nowhere else to go and lands directly on the stick coordinate:
  `coordinates[-1] = d0 + 2`.
- `_multi_arg_pointwise_layouts`'s `_is_supported_layout`
  (`propagate_layouts.py:1701`) requires every arg's stick coordinate to
  satisfy `is_stick_expr_offset_free` (a bare symbol, `Mod(x, N)`, or `0`)
  for every candidate output stick dim. `d0 + 2` fails this check for both
  candidates tried (stick on output dim 0 and dim 1), so no output layout
  survives and `Unsupported` is raised.

Confirmed with direct A/B instrumentation:
- Offset 0 (`case_a_offset0`): `ids`'s stick coordinate is the bare symbol
  `d0` -> passes `is_stick_expr_offset_free`.
- Offset 2 (`case_a_offset2`, this issue's A2): stick coordinate is `d0 + 2`
  -> fails.
- Not a "too-small tensor" artifact: scaling `ids` up to 128 elements (2
  full 64-element sticks, `device_size=[2, 64]`) reproduces the identical
  `d0 + 2` coordinate and identical failure -- the outer dim still has
  `size=2 > 1` in that case, yet `add_term`'s per-dim modular split (the
  fallback `compute_coordinates` uses when `_decompose_constant_offset`
  isn't reached, or here, when it *does* land at dim 1 successfully but
  dim 1 is the stick dim) still resolves the offset onto the stick
  coordinate rather than splitting `d0 + 2` into
  `[floor((d0+2)/64), Mod(d0+2, 64)]`. So the real gap is that no code path
  attempts a floor/Mod split of an offset landing on the stick dimension
  itself, even when doing so would trivially keep the stick coordinate
  offset-free (a `Mod` is on `is_stick_expr_offset_free`'s allow-list).

## Assessment

This is an architectural gap, not a narrow bug: no mechanism currently
exists to either (a) commit `ids` to a device layout that reserves a
splittable outer dimension when a downstream `narrow` is present, or
(b) split an offset stick coordinate into `floor`/`Mod` when it lands on
the stick dim, the way `add_term`'s ordinary per-dim splitting already
does for loop-variable terms (`views.py:341` / `:349`,
`var * step % next_stride[dim] // st`). `_is_stick_expr_with_offset`
(`pass_utils.py:1363`) and its caller `_check_stick_expr_supported`
already document this shape as "may be valid but is not yet supported" --
this issue is that unimplemented case, reached through
`_multi_arg_pointwise_layouts`'s stricter `is_stick_expr_offset_free`
check rather than through `_check_stick_expr_supported`'s path.

## Repro / instrumentation used

Traced by monkey-patching `try_device_coordinates`, `is_stick_expr_offset_free`,
`_compute_dim_order`, and `_decompose_constant_offset` around the issue's
own `case_a_offset2` reproducer. Happy to attach the instrumentation script
if useful for whoever picks up the fix.

**Comment by dgrove-oss** [comment]
Removed phase-1 label, spyre-inference team confirmed this is non-blocking.
